backtrace_test.go

```diff
@@ -1,88 +1,185 @@
+			expected: `k8s.io/klog/v2_test.TestBacktrace.funcX():
+zzz/backtrace_test.go:xxx
+k8s.io/klog/v2_test.inner():
+zzz/backtrace_test.go:47
+k8s.io/klog/v2_test.outer():
+zzz/backtrace_test.go:43
+k8s.io/klog/v2_test.TestBacktrace.funcX():
+zzz/backtrace_test.go:xxx
+`,
+		},
+
+		"skip1": {
+			callback: func() interface{} { return klog.Backtrace(klog.BacktraceSkip(1)) },
+			expected: `k8s.io/klog/v2_test.inner():
+zzz/backtrace_test.go:47
+k8s.io/klog/v2_test.outer():
+zzz/backtrace_test.go:43
+k8s.io/klog/v2_test.TestBacktrace.funcX():
+zzz/backtrace_test.go:xxx
+`,
+		},
 
-	//anonymous functions
-	c := func() (bool, string) {
-		s := fmt.Sprintf("%v", Backtrace())
-		p1 := regexp.MustCompile("(?s)func1\\(\\).*func2\\(\\).*func3\\(\\)")
-		return p1.MatchString(s), s
+		"skip2": {
+			callback: func() interface{} { return klog.Backtrace(klog.BacktraceSkip(2)) },
+			expected: `k8s.io/klog/v2_test.outer():
+zzz/backtrace_test.go:43
+k8s.io/klog/v2_test.TestBacktrace.funcX():
+zzz/backtrace_test.go:xxx
+`,
+		},
+
+		"trace1": {
+			callback: func() interface{} { return klog.Backtrace(klog.BacktraceSize(1)) },
+			expected: `k8s.io/klog/v2_test.TestBacktrace.funcX():
+zzz/backtrace_test.go:xxx
+`,
+		},
+
+		"trace2": {
+			callback: func() interface{} { return klog.Backtrace(klog.BacktraceSize(2)) },
+			expected: `k8s.io/klog/v2_test.TestBacktrace.funcX():
+zzz/backtrace_test.go:xxx
+k8s.io/klog/v2_test.inner():
+zzz/backtrace_test.go:47
+`,
+		},
+
+		"skip1-trace1": {
+			callback: func() interface{} { return klog.Backtrace(klog.BacktraceSize(1), klog.BacktraceSkip(1)) },
+			expected: `k8s.io/klog/v2_test.inner():
+zzz/backtrace_test.go:47
+`,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			backtrace := outer(tt.callback)
+			actual := normalizeBacktrace(t, backtrace)
+			if actual != tt.expected {
+				t.Errorf("Wrong backtrace. Expected:\n%s\nActual:\n%s\n", tt.expected, actual)
+			}
+		})
 	}
-	b := func() (bool, string) { return c() }
-	a := func() (bool, string) { return b() }
-	ret, str := a()
-	if ret == false {
-		t.Errorf("Expected to see 3 anonymous functions (func1/func2/func3) got %v", str)
+}
+
+func normalizeBacktrace(t *testing.T, backtrace interface{}) string {
+	str := backtrace.(fmt.Stringer).String()
+
+	// This matches the stack entry for the testing package:
+	// testing.tRunner():
+	//     /nvme/gopath/go-1.18.1/src/testing/testing.go:1439
+	//
+	// It and all following entries vary and get removed.
+	end := regexp.MustCompile(`(?m)^testing\.`).FindStringIndex(str)
+	if end != nil {
+		str = str[:end[0]]
 	}
 
+	// Remove varying line numbers.
+	str = regexp.MustCompile(`(?m):([67890][0-9]|[1-9][0-9][0-9])$`).ReplaceAllString(str, ":xxx")
+
+	// Remove varying anonymous function numbers.
+	str = regexp.MustCompile(`\.func[[:digit:]]+`).ReplaceAllString(str, ".funcX")
+
+	// Remove varying path names
+	str = regexp.MustCompile(`(.*)(backtrace_test.go:.*)`).ReplaceAllString(str, "zzz/$2")
```

Comment:
Let's keep the whitespace in front of the path name. It makes the expected output closer to the original and easier to read.

```suggestion
	str = regexp.MustCompile(`([[:blank:]]+)[[:^blank:]]*(backtrace_test.go:.*)`).ReplaceAllString(str, "$1/zzz/$2")
```